src/oauth2/service.ts

```diff
@@ -169,6 +171,33 @@ export async function generateTokenFromCode(client: OAuth2Client, code: string):
 
 }
 
+export function validatePKCE(codeVerifier: string|undefined, codeChallenge: string|undefined, codeChallengeMethod: string|undefined) {
+  if (!codeChallenge && !codeVerifier) {
+    // This request was not initiated with PKCE support, so ignore the validation
+    return;
+  }
+
+  if (codeChallenge && !codeVerifier) {
+    // The authorization request started with PKCE, but the token request did not follow through
+    throw new InvalidRequest('The code verifier was not supplied');
+  }
+
+  // For the plain method, the derived code and the code verifier are the same
+  let derivedCodeChallenge = codeVerifier;
+
+  if (codeChallengeMethod === 'S256') {
```

Comment:
This is a suggestion, not a must... but if this were to be done with a switch (that also handles `plain`) you could use this approach:

https://stackoverflow.com/a/58009992/80911

to ensure that Typescript always complains if the enum ever gets expanded with a third "method".

The switch with `never` type is a bit ugly, so another option is to move this to a function that takes a `codeChallengeMethod`, a `codeVerifier` and returns the `derivedCodeChallenge`.

Reply:
I feel like a switch statement is overkill for only two possibilities. If there was a third type of challenge method, I'd create a separate function. I guess I could have done this with a ternary, too.

Reply:
The suggestion has more to related to getting a static error check once a third option is added. Basically when `CodeChallengeMethod` is altered with a new option, in the ideal case typescript refuses to compile it until that new case is explicitly handled. Switch is one way to handle this.. but this is not a big deal.

BTW I know other code in this repo is _also_ not really guarded as well and written this way.. I changed how I write things quite a bit since I wrote this stuff =)

Anyway, this works!